runtime/caml/eventring.h

```diff
@@ -0,0 +1,308 @@
+    struct caml_eventring_cursor *cursor,
+    void *callback_data,
+    uintnat max_events, uintnat *events_consumed);
+
+/* OCaml API for reading from the eventring. Documented in eventring.ml */
+extern value caml_eventring_create_cursor_ml(value path_pid);
+extern value caml_eventring_free_cursor_ml(value wrapped_cursor);
+extern value caml_eventring_read_poll_ml(value wrapped_cursor,
+                                              value callbacks,
+                                              value max_events_option);
+
+#ifdef CAML_INTERNALS
+
+struct eventring_buffer_header {
+  atomic_uint_fast64_t ring_head;
+  atomic_uint_fast64_t ring_tail;
+  uint64_t padding[8]; /* Padding so headers don't share cache lines. Eight
+                          words guarantees that buffer headers don't share
+                          cache lines, even for non-aligned allocations. */
+};
+
+/* For a more detailed explanation of the eventring file layout, see
+   eventring.c */
+struct eventring_metadata_header {
+  uint64_t version;
+  uint64_t max_domains;
+  uint64_t ring_header_size_bytes; /* Ring buffer header size (bytes) */
+  uint64_t ring_size_bytes; /* Ring data size (bytes) */
+  uint64_t ring_size_elements; /* Ring size in 64-bit elements */
+  uint64_t headers_offset; /* Offset from this struct to first header (bytes) */
+  uint64_t data_offset; /* Offset from this struct to first data (byte) */
+  uint64_t padding; /* Make the header a multiple of 64 bytes */
+};
+
+#define EVENTRING_NUM_ALLOC_BUCKETS 20
+#define EVENTRING_MAX_MSG_LENGTH (1 << 10)
+
+/* event header fields (for runtime events):
+| -- length (10 bits) -- | runtime or user event (1 bit) | event type (4 bits) |
+event id (13 bits)
+*/
+
+#define EVENTRING_ITEM_LENGTH(header) (((header) >> 54) & ((1UL << 10) - 1))
+#define EVENTRING_ITEM_IS_RUNTIME(header) !((header) | (1UL << 53))
+#define EVENTRING_ITEM_IS_USER(header) ((header) | (1UL << 53))
+#define EVENTRING_ITEM_TYPE(header) (((header) >> 49) & ((1UL << 4) - 1))
+#define EVENTRING_ITEM_ID(header) (((header) >> 36) & ((1UL << 13) - 1))
+
+/* Set up eventring (and check if we need to start it immediately). Called
+   from startup* */
+void caml_eventring_init();
+
+/* Destroy all allocated eventring structures and clear up the ring. Called
+   from [caml_sys_exit] */
+void caml_eventring_destroy();
+
+/* Handle safely re-initialising the eventring structures in a forked child */
+void caml_eventring_post_fork();
+
+/* Returns the location of the eventring for the current process if started or
+   NULL otherwise */
+char_os* caml_eventring_current_location();
+
+/* Functions for putting runtime data on to the eventring */
+void caml_ev_begin(ev_runtime_phase phase);
+void caml_ev_end(ev_runtime_phase phase);
+void caml_ev_counter(ev_runtime_counter counter, uint64_t val);
+void caml_ev_lifecycle(ev_lifecycle lifecycle, int64_t data);
+void caml_ev_alloc(uint64_t sz);
+void caml_ev_alloc_flush();
+void caml_ev_flush();
```

Comment:
The behavior of `ev_alloc`, `ev_alloc_flush` and `ev_flush` cannot be guessed (by met at least) from the API. (eventring.c has some documentation comments, but I would rather expect that the .h file lets me know how to use the API).

Bonus question: do we really need `caml_ev_flush`? It seems to be a no-op.

Reply:
So `ev_alloc`, `ev_alloc_flush` and `ev_flush` are part of the existing eventlog API. The former two are only used in specific places and track the size of allocations in the major heap.

In https://github.com/ocaml/ocaml/pull/10964/commits/1f55df7cde62b4830e7bd4ee009523bc0c1f7315 I've removed `caml_ev_flush` as well as moved to eventring.h and expanded the `caml_ev_alloc` comment from eventring.c.